cardano-crypto-class/src/Cardano/Crypto/KES/CompactSum.hs

```diff
@@ -0,0 +1,400 @@
+      mconcat
+        [ rawSerialiseSignKeyKES sk
+        , getSeedBytes r_1
+        , rawSerialiseVerKeyKES vk_0
+        , rawSerialiseVerKeyKES vk_1
+        ]
+
+    rawSerialiseSigKES (SigCompactSumKES sigma vk_other) =
+      mconcat
+        [ rawSerialiseSigKES sigma
+        , rawSerialiseVerKeyKES vk_other
+        ]
+
+    rawDeserialiseVerKeyKES = fmap VerKeyCompactSumKES  . hashFromBytes
+
+    rawDeserialiseSignKeyKES b = do
+        guard (BS.length b == fromIntegral size_total)
+        sk   <- rawDeserialiseSignKeyKES b_sk
+        let r = mkSeedFromBytes          b_r
+        vk_0 <- rawDeserialiseVerKeyKES  b_vk0
+        vk_1 <- rawDeserialiseVerKeyKES  b_vk1
+        return (SignKeyCompactSumKES sk r vk_0 vk_1)
+      where
+        b_sk  = slice off_sk  size_sk b
+        b_r   = slice off_r   size_r  b
+        b_vk0 = slice off_vk0 size_vk b
+        b_vk1 = slice off_vk1 size_vk b
+
+        size_sk    = sizeSignKeyKES (Proxy :: Proxy d)
+        size_r     = seedSizeKES    (Proxy :: Proxy d)
+        size_vk    = sizeVerKeyKES  (Proxy :: Proxy d)
+        size_total = sizeSignKeyKES (Proxy :: Proxy (CompactSumKES h d))
+
+        off_sk     = 0 :: Word
+        off_r      = size_sk
+        off_vk0    = off_r + size_r
+        off_vk1    = off_vk0 + size_vk
+
+    rawDeserialiseSigKES b = do
+        guard (BS.length b == fromIntegral size_total)
+        sigma <- rawDeserialiseSigKES    b_sig
+        vk  <- rawDeserialiseVerKeyKES b_vk
+        return (SigCompactSumKES sigma vk)
+      where
+        b_sig = slice off_sig size_sig b
+        b_vk  = slice off_vk  size_vk  b
+
+        size_sig   = sizeSigKES    (Proxy :: Proxy d)
+        size_vk    = sizeVerKeyKES (Proxy :: Proxy d)
+        size_total = sizeSigKES    (Proxy :: Proxy (CompactSumKES h d))
+
+        off_sig    = 0 :: Word
+        off_vk     = size_sig
+
+instance (KESAlgorithm (CompactSumKES h d), OptimizedKESAlgorithm d, HashAlgorithm h) => OptimizedKESAlgorithm (CompactSumKES h d) where
+    verifySigKES ctxt t a (SigCompactSumKES sigma _) =
+      verifySigKES ctxt t' a sigma
+      where
+        _T = totalPeriodsKES (Proxy :: Proxy d)
+        t' | t < _T = t
+           | otherwise = t - _T
+
+    verKeyFromSigKES ctxt t (SigCompactSumKES sigma vk_other) =
+      VerKeyCompactSumKES $ hashPairOfVKeys (vk_0, vk_1)
+      where
+        _T = totalPeriodsKES (Proxy :: Proxy d)
+        t' | t < _T = t
+           | otherwise = t - _T
+        (vk_0, vk_1) | t < _T = (verKeyFromSigKES ctxt t' sigma, vk_other)
+                     | otherwise = (vk_other, verKeyFromSigKES ctxt t' sigma)
+
+hashPairOfVKeys :: (KESAlgorithm d, HashAlgorithm h)
+                => (VerKeyKES d, VerKeyKES d)
+                -> Hash h (VerKeyKES d, VerKeyKES d)
+hashPairOfVKeys =
+    hashWith $ \(a,b) ->
+      rawSerialiseVerKeyKES a <> rawSerialiseVerKeyKES b
+
+slice :: Word -> Word -> ByteString -> ByteString
```

Comment:
I feel like we should move these into a shared module, rather than repeating them all over the place.

Reply:
Agree. Will do.